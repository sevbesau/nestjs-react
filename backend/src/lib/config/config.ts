import { config } from 'dotenv';
import { join } from 'path';
import { z, ZodError } from 'zod';

enum Environment {
  DEVELOPMENT = 'development',
  TEST = 'test',
  PRODUCTION = 'production',
}

const envExtensions = {
  [Environment.DEVELOPMENT]: 'local',
  [Environment.TEST]: 'test',
  [Environment.PRODUCTION]: '',
};

export const environment =
  (process.env.NODE_ENV as Environment) || Environment.DEVELOPMENT;
export const isProductionEnvironment = environment === Environment.PRODUCTION;
export const isTestEnvironment = environment === Environment.TEST;
export const isDevelopmentEnvironment = environment === Environment.DEVELOPMENT;

if (!isProductionEnvironment) {
  config({
    path: join(__dirname, `../../../.env.${envExtensions[environment]}`),
  });
}

const ConfigZodSchema = z.object({
  PORT: z.coerce.number(),
  BASE_URL: z.string(),

  ALLOWED_ORIGINS: z.string(),

  API_NAME: z.string(),
  API_DESCRIPTION: z.string(),
  API_VERSION: z.string(),

  DOCUMENTATION_PATH: z.string(),
  DOCUMENTATION_USER: z.string(),
  DOCUMENTATION_PASSWORD: z.string(),

  DATABASE_URI: z.string(),
  DATABASE_NAME: z.string(),

  COOKIE_SECRET: z.string(),

  JWT_ACCESS_TOKEN_SECRET: z.string(),
  JWT_ACCESS_TOKEN_EXPIRATION: z.string().default('1h'),
  JWT_REFRESH_TOKEN_SECRET: z.string(),
  JWT_REFRESH_TOKEN_EXPIRATION: z.string().default('1d'),

  CAPTCHA_PRIVATE_KEY: z.string(),
  BYPASS_CAPTCHA: z.coerce.boolean(),

  SENDGRID_KEY: z.string(),

  MAILS_FROM_NAME: z.string(),
  MAILS_FROM_EMAIL: z.string().email(),

  CONTACT_REQUEST_INBOX: z.string().email(),
});

export type TConfig = z.infer<typeof ConfigZodSchema>;

export default () => {
  try {
    return ConfigZodSchema.parse(process.env);
  } catch (error) {
    const missingVars = (error as ZodError).errors.reduce(
      (missing: string[], error) => {
        return [...missing, error.path];
      },
      [],
    );
    throw new Error(`Missing env vars:\n  ${missingVars.join('\n  ')}`);
  }
};
