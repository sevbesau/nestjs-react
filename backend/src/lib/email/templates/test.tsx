import { Body, Head, Html, Preview } from '@react-email/components';
import * as React from 'react';

import { EmailTemplate, EmailTemplateContext } from '.';

interface TestTemplateProps {
  hello: string;
}

export const testTemplate = new EmailTemplate(
  () => 'subject',
  ({ hello }: TestTemplateProps, {}: EmailTemplateContext) => {
    return (
      <Html>
        <Head />
        <Preview>{hello}</Preview>
        <Body>{hello}</Body>
      </Html>
    );
  },
);

export default testTemplate;
