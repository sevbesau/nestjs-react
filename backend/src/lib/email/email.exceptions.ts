import { isProductionEnvironment } from '@lib/config/config';
import { InternalServerErrorException } from '@nestjs/common';

export class FailedToSendEmailException extends InternalServerErrorException {
  constructor() {
    super(
      isProductionEnvironment
        ? 'Something went wrong...'
        : 'Failed to send email using sendgrid',
    );
  }
}
