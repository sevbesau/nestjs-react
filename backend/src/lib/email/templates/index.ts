import { TConfig } from '@lib/config/config';
import { I18nTranslations } from '@lib/i18n/generated';
import { ConfigService } from '@nestjs/config';
import { I18nService } from 'nestjs-i18n';

export interface EmailTemplateContext {
  i18n: I18nService<I18nTranslations>;
  configService: ConfigService<TConfig>;
  lang: string;
}

export class EmailTemplate<T> {
  constructor(
    readonly subject: (context: EmailTemplateContext) => string,
    readonly component: (
      props: T,
      context?: EmailTemplateContext,
    ) => JSX.Element,
  ) {}
}

export type PropsOfTemplate<T extends EmailTemplate<unknown>> = Parameters<
  T['component']
>[0];
