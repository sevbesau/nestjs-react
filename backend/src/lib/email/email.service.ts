import { TConfig } from '@lib/config/config';
import { I18nTranslations } from '@lib/i18n/generated';
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { render } from '@react-email/render';
import sendgridService, { MailDataRequired } from '@sendgrid/mail';
import { exec } from 'child_process';
import { writeFile } from 'fs';
import { I18nService } from 'nestjs-i18n';
import { file } from 'tmp';

import { FailedToSendEmailException } from './email.exceptions';
import {
  EmailTemplate,
  EmailTemplateContext,
  PropsOfTemplate,
} from './templates';

@Injectable()
export class EmailService {
  constructor(
    private readonly configService: ConfigService<TConfig>,
    private readonly i18n: I18nService<I18nTranslations>,
  ) {
    // Useful for developing emails and opening a preview in the browser
    //     const { html } = this.renderTemplate(contactRequestNotificationTemplate, {
    //       email: 'seppe@siliconminds.be',
    //       firstName: 'seppe',
    //       lastName: 'van besauw',
    //       message: `hello world,
    // this is a beautiful day.
    // How are you?
    // Best regards.`,
    //     });
    // this.openHtmlInBrowser(html);
  }

  async onModuleInit() {
    sendgridService.setApiKey(this.configService.getOrThrow('SENDGRID_KEY'));
  }

  async sendTemplateMail<T extends EmailTemplate<unknown>>(
    template: T,
    props: PropsOfTemplate<T>,
    to: string | string[],
    options?: { lang?: string },
  ) {
    const { html, text, subject } = this.renderTemplate(
      template,
      props,
      options?.lang,
    );
    return this.sendEmail(to, subject, text, html);
  }

  private async sendEmail(
    to: string | string[],
    subject: string,
    text: string,
    html: string,
  ) {
    const msg: MailDataRequired = {
      to,
      from: {
        name: this.configService.getOrThrow('MAILS_FROM_NAME'),
        email: this.configService.getOrThrow('MAILS_FROM_EMAIL'),
      },
      subject,
      text,
      html,
    };
    try {
      await sendgridService.send(msg);
    } catch (error) {
      throw new FailedToSendEmailException();
    }
  }

  private renderTemplate<T extends EmailTemplate<unknown>>(
    template: T,
    props: PropsOfTemplate<T>,
    lang?: string,
  ) {
    const context = this.createEmailContext(lang);
    const componentWithProps = template.component(props, context);
    const html = render(componentWithProps, { pretty: true });
    const text = render(componentWithProps, { plainText: true });
    const subject = template.subject(context);

    return { html, text, subject };
  }

  private createEmailContext(lang = 'nl'): EmailTemplateContext {
    return {
      i18n: this.i18n,
      configService: this.configService,
      lang,
    };
  }

  private openHtmlInBrowser(html: string) {
    file((err, path, _fd, cleanupCallback) => {
      if (err) throw err;
      writeFile(path, html, (err) => {
        if (err) throw err;
        this.openUrl(`file://${path}`);
      });
      setTimeout(cleanupCallback, 10_000);
    });
  }

  private openUrl(url: string) {
    const openUrl =
      process.platform == 'darwin'
        ? 'open'
        : process.platform == 'win32'
        ? 'start'
        : 'xdg-open';
    exec(`${openUrl} ${url}`);
  }
}
