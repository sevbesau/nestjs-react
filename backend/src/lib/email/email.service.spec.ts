import { createMockConfigService } from '@lib/test/config';
import { ConfigService } from '@nestjs/config';
import { Test, TestingModule } from '@nestjs/testing';
import sendgridService from '@sendgrid/mail';
import { I18nService } from 'nestjs-i18n';
import { instance, mock } from 'ts-mockito';

import { FailedToSendEmailException } from './email.exceptions';
import { EmailService } from './email.service';
import testTemplate from './templates/test';

describe('EmailService', () => {
  let emailService: EmailService;
  const mockConfigService = createMockConfigService({
    SENDGRID_KEY: 'key',
    MAILS_FROM_NAME: 'test',
    MAILS_FROM_EMAIL: 'test@tester.com',
  });
  const mockI18N = mock(I18nService);

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [EmailService],
    })
      .useMocker((token) => {
        if (token === ConfigService) return instance(mockConfigService);
        if (token === I18nService) return instance(mockI18N);
      })
      .compile();

    emailService = module.get<EmailService>(EmailService);
  });

  it('should be defined', () => {
    expect(emailService).toBeDefined();
  });

  describe('sendTemplateMail', () => {
    it('should be defined', () => {
      expect(emailService.sendTemplateMail).toBeDefined();
    });

    it('should render and send an email', async () => {
      const sendEmailSpy = jest
        .spyOn(emailService as any, 'sendEmail')
        .mockResolvedValue(null);
      const mockEmail = {
        subject: 'subject',
        html: 'html',
        text: 'text',
      };
      const renderTemplateSpy = jest
        .spyOn(emailService as any, 'renderTemplate')
        .mockReturnValue(mockEmail);
      const props = { hello: 'world' };
      const recipient = 'fake@email.com';

      await emailService.sendTemplateMail(testTemplate, props, recipient);

      expect(renderTemplateSpy).toBeCalledWith(testTemplate, props, undefined);
      expect(sendEmailSpy).toBeCalledWith(
        recipient,
        mockEmail.subject,
        mockEmail.text,
        mockEmail.html,
      );
      jest.spyOn(emailService as any, 'sendEmail').mockRestore();
      jest.spyOn(emailService as any, 'renderTemplate').mockRestore();
    });
  });

  describe('sendEmail', () => {
    it('should be defined', () => {
      expect(emailService['sendEmail']).toBeDefined();
    });

    it('should send an email using the sendgrid api', async () => {
      const sendGridSpy = jest
        .spyOn(sendgridService, 'send')
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        .mockResolvedValue(null);
      const mockSubject = 'subject';
      const mockHtml = 'html';
      const mockText = 'text';
      const mockRecipient = 'fake@email.com';

      await emailService['sendEmail'](
        mockRecipient,
        mockSubject,
        mockText,
        mockHtml,
      );

      const msg = sendGridSpy.mock.lastCall;

      expect(msg).toMatchSnapshot();
      jest.spyOn(sendgridService, 'send').mockRestore();
    });

    it('should throw when sending fails', async () => {
      jest.spyOn(sendgridService, 'send').mockRejectedValue(new Error());
      const mockSubject = 'subject';
      const mockHtml = 'html';
      const mockText = 'text';
      const mockRecipient = 'fake@email.com';

      await expect(
        emailService['sendEmail'](
          mockRecipient,
          mockSubject,
          mockText,
          mockHtml,
        ),
      ).rejects.toThrow(FailedToSendEmailException);

      jest.spyOn(sendgridService, 'send').mockRestore();
    });
  });

  describe('renderTemplate', () => {
    it('should be defined', () => {
      expect(emailService['renderTemplate']).toBeDefined();
    });

    it('should render a template and return html, text and the subject', () => {
      const { html, text, subject } = emailService['renderTemplate'](
        testTemplate,
        { hello: 'world' },
      );

      expect(subject).toBe('subject');
      expect(text).toBe('world');
      expect(html).toMatchSnapshot();
    });
  });

  describe('createEmailContext', () => {
    it('should be defined', () => {
      expect(emailService['createEmailContext']).toBeDefined();
    });

    it('should create a context object for the email template class', () => {
      const context = emailService['createEmailContext']();

      expect(context.i18n).toBeDefined();
      expect(context.configService).toBeDefined();
      expect(context.lang).toBeDefined();
    });
  });
});
