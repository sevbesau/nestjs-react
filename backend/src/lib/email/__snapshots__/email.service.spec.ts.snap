// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailService renderTemplate should render a template and return html, text and the subject 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html lang="en">

  <head data-id="__react-email-head">
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
  </head>
  <div id="__react-email-preview" style="display:none;overflow:hidden;line-height:1px;opacity:0;max-height:0;max-width:0">world<div> ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿</div>
  </div>

  <body data-id="__react-email-body">world</body>

</html>"
`;

exports[`EmailService sendEmail should send an email using the sendgrid api 1`] = `
[
  {
    "from": {
      "email": "test@tester.com",
      "name": "test",
    },
    "html": "html",
    "subject": "subject",
    "text": "text",
    "to": "fake@email.com",
  },
]
`;
